### PR TITLE
[Use Cases] Add use cases for mounting

### PR DESCRIPTION
**Pull request by flo91**
## Basics

- [x] Short descriptions of your changes are in the release notes
      (added as entry in `doc/news/_preparation_next_release.md` which
      contains `_(my name)_`)
      **Please always add something to the release notes.**
- [x] Details of what you changed are in commit messages
      (first line should have `module: short statement` syntax)
- [ ] References to issues, e.g. `close #X`, are in the commit messages.
- [ ] The buildservers are happy. If not, fix **in this order**:
  - [ ] add a line in `doc/news/_preparation_next_release.md`
  - [ ] reformat the code with `scripts/dev/reformat-all`
  - [ ] make all unit tests pass
  - [ ] fix all memleaks
- [x] The PR is rebased with current master.

## Checklist

- [ ] I added unit tests for my code
- [ ] I fully described what my PR does in the documentation
      (not in the PR description)
- [ ] I fixed all affected documentation (see [Documentation Guidelines](https://www.libelektra.org/devgettingstarted/documentation))
- [ ] I fixed all affected decisions (see [Decision Process](https://www.libelektra.org/decisions/decision-process))
- [ ] I added code comments, logging, and assertions as appropriate (see [Coding Guidelines](https://www.libelektra.org/devgettingstarted/coding))
- [ ] I updated all meta data (e.g. README.md of plugins and [METADATA.ini](https://master.libelektra.org/doc/METADATA.ini))
- [ ] I mentioned [every code](/.reuse/dep5) not directly written by me in [reuse syntax](https://reuse.software/)

## Review

- [ ] Documentation is introductory, concise, good to read and describes everything what the PR does
- [ ] Examples are well chosen and understandable
- [ ] Code is conforming to [our Coding Guidelines](https://master.libelektra.org/doc/CODING.md)
- [ ] APIs are conforming to [our Design Guidelines](https://master.libelektra.org/doc/DESIGN.md)
- [ ] Code is consistent to [our Design Decisions](https://master.libelektra.org/doc/decisions)

## Labels

- [ ] Add the "work in progress" label if you do not want the PR to be reviewed yet.
- [ ] Add the "ready to merge" label **if the basics are fulfilled** and no further pushes are planned by you.

### CODE REVIEW

File: doc/usecases/mount/UC_mount.md [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+- **Main success scenario:**
+
+  - The administrators specify a mountpoint in the KDB and an identifier for the data source they want to mount.
+    - For file-based data sources, the identifier is usually the path and filename.
+  - The data source is successfully mounted into the KDB.
+
+- **Alternative scenario:**
+
+  - The given data source is not found. (e.g. wrong filename)
+    - Elektra reports the error to the administrators.
```
Comment (markus2330) [inline doc/usecases/mount/UC_mount.md:26]: There is no need that a mounted config file needs to be present. What is an alternative scenario, though is:
```suggestion
  - There are already keys present at the requested mountpoint.
    - Elektra reports the error to the administrators, unless `--strategy` is used, then keys are merged as requested.
```

File: doc/usecases/mount/UC_mount.md [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+  - The given data source is not supported by Elektra. (e.g. wrong file format, syntax errors)
+    - Elektra reports the error to the administrators.
```
Comment (markus2330) [inline doc/usecases/mount/UC_mount.md:28]: Why isn't this an error scenario?

File: doc/usecases/mount/UC_mount.md [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+- **Error scenario:**
+
+  - The opening of the data source failed.
+    - Elektra reports the error to the administrators.
```
Comment (markus2330) [inline doc/usecases/mount/UC_mount.md:33]: Please describe what "opening of the data source failed" means. I don't think this is done during mounting except for the `--strategy` described by me above.

File: doc/usecases/mount/UC_mount.md [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+- **Postconditions:**
+
+  - The newly mounted data source is listed when querying the current mountpoints. (see [Use Case: List all mountpoints](./UC_list_mountpoints.md))
+  - The administrators can now work (CRUD) with the configuration data stored in the newly mounted data source using Elektras key-value-based interface. (see other mounting use cases)
+
+- **Non-functional Constraints:**
+
+  - The mounting of new data sources should not take longer than 5 seconds (reference system has to be defined).
```
Comment (markus2330) [inline doc/usecases/mount/UC_mount.md:42]:
```suggestion
  - Mounting should be reasonable fast for interactive use.
```

File: doc/usecases/mount/UC_unmount.md [hunk 5]
```diff
@@ -0,0 +1,47 @@
+# Use Case: Unmount a data source
+
+## Summary
+
+- **Scope:** Mounting
+- **Level:** Kite
+- **Actors:** Administrators
```
Comment (markus2330) [inline doc/usecases/mount/UC_unmount.md:7]: Its also applications, that's why we need a library.

File: doc/usecases/mount/UC_unmount.md [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+
+  - The given mountpoint is not found.
+    - Elektra reports the errors to the administrators.
+
+- **Error scenario:**
+
+  - The opening of the data source failed. (e.g. the mounted file is no longer available)
+    - Elektra reports the errors to the administrators.
+  - The unmounting is not possible. (e.g. because the data source is currently in use by Elektra)
+    - Elektra reports the errors to the administrators.
```
Comment (markus2330) [inline doc/usecases/mount/UC_unmount.md:37]: Do these error cases really exist? Imho only writing to the mountpoint configuration could fail?

File: doc/usecases/mount/UC_list_mountpoints.md [hunk 7]
```diff
@@ -0,0 +1,47 @@
+# Use Case: List all mountpoints
+
+## Summary
+
+- **Scope:** Mounting
+- **Level:** Kite
+- **Actors:** Administrators
+- **Brief:** The administrators can query the current mountpoints.
+- **Status:** Implemented (currently only file-based data sources can be mounted)
```
Comment (markus2330) [inline doc/usecases/mount/UC_list_mountpoints.md:9]:
```suggestion
- **Status:** Draft (currently only file-based data sources can be listed)
```

File: doc/usecases/mount/UC_list_mountpoints.md [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+- **Main success scenario:**
+
+  - The administrators use Elektra to query a list with all current mountpoints.
+  - Elektra returns the list with the mountpoints and the data source identifiers. (e.g. filename and path, data source name)
+  - The administrators mount a new data source. (see [Use Case: Mount a data source](./UC_mount.md))
+  - The administrators use Elektra to query a list with all current mountpoints.
+  - The new mountpoint is now added to the list.
+  - The administrators unmount the previously mounted data source. (see [Use Case: Unmount a data source](./UC_unmount.md))
+  - The administrators use Elektra to query a list with all current mountpoints.
+  - The mountpoint for the unmounted data source is now not included in the list.
```
Comment (markus2330) [inline doc/usecases/mount/UC_list_mountpoints.md:26]: The other use cases already cover mounting/umounting:

```suggestion
  - The administrators use Elektra to query a list of all current mountpoints.
```

File: doc/usecases/mount/UC_list_mountpoints.md [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+
+  - none new
+  - The scenarios of the used use cases (mounting and unmounting) apply.
+
+- **Error scenario:**
+
+  - An error occurred while mounting the data source. (see [Use Case: Mount a data source](./UC_mount.md))
+    - Elektra reports the error to the administrators.
+  - An error occurred while unmounting the data source. (see [Use Case: Unmount a data source](./UC_unmount.md))
+    - Elektra reports the error to the administrators.
```
Comment (markus2330) [inline doc/usecases/mount/UC_list_mountpoints.md:38]: I don't think we need an error scenario here.

```suggestion
```

File: doc/usecases/mount/UC_list_values.md [hunk 10]
```diff
@@ -0,0 +1,40 @@
+# Use Case: List all configuration keys and values from all currently mounted data sources
```
Comment (markus2330) [inline doc/usecases/mount/UC_list_values.md:1]: I don't know which functionality this is? Is this `export`? Why does it belong to the mounting use cases?

File: doc/usecases/mount/UC_delete_from_file.md [hunk 11]
```diff
@@ -0,0 +1,52 @@
+# Use Case: Delete configuration data from a file
```
Comment (markus2330) [inline doc/usecases/mount/UC_delete_from_file.md:1]: I would simplify this and only have a get and set use case. (Delete is part of set).
I think it is enough that you only write the rdb use cases.
I don't think they should be part of `mount`, but rather another folder ("rdb").

File: doc/usecases/mount/UC_edit_rdb.md [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+  - The administrators use Elektra to change the value of the given key, by giving the unique keyname and the new value.
+  - Elektra updates the affected tuple in the database.
+  - The administrators query the same key as in the first step.
+  - Elektra returns the new value of the given key to the administrators.
+  - The administrators query the database directly to verify that the requested change was written.
+
+- **Alternative scenario:**
+
+  - The requested key is not found in the database.
+    - Elektra reports the error to the administrators.
```
Comment (markus2330) [inline doc/usecases/mount/UC_edit_rdb.md:31]: As Elektra simply `set` whatever is given, the non-presence of keys shouldn't matter (they get created/deleted as demanded).

**Review by markus2330 — CHANGES_REQUESTED** [review]
Thank you for the use cases! A few things I didn't understand. I didn't review all file/rdb use cases as they seem to be repetitive, please re-request review if I missed something.

**Review by kodebach — COMMENTED** [review]
I haven't had time to read all the texts, but from the titles and briefs, I'd say only these three use cases are use cases for mounting: `UC_mount.md`, `UC_unmount.md` and `UC_list_mountpoints.md`

The rest are use cases for file-based backends and database backends respectively. If we want to keep them I'd move them into separate folders. However, I'm not sure we need those at all. 

IMO it would be enough to have a `UC_mount_file.md` and `UC_mount_rdb.md` in addition to `UC_mount.md`. `UC_mount.md` describes mounting in general, and the other two describe mounting configuration files and relational databases respectively. Then all the edit/read/write/delete stuff should automatically follow as a combination of  `UC_mount_file.md` and `UC_mount_rdb.md` and the general KDB use cases in `doc/usecases/kdb`.

Only if there are specific difference between file-based and database backends that go beyond technical limitations (e.g. files must be written as a whole, databases could modify individual values), would I create separate use cases. The technical limitations are only relevant on the decision level IMO, but not the more general use case level.

File: doc/usecases/mount/UC_mount.md (hunk 1 above)
Comment (kodebach) [reply doc/usecases/mount/UC_mount.md:26]: I'm not sure if this what you meant, but your suggestion sounds like: We tried e.g. `kdb mount foo user:/foo`, but keys below `user:/foo` already existed (without a mountpoint at `user:/foo`). If you meant, there was already a mountpoint at `user:/foo`, I'd put it like this for more clarity:

```suggestion
  - The desired mountpoint already exists with a different configuration.
    - Elektra reports the error to the administrators, unless `--strategy` is used, then keys are merged as requested.
```

File: doc/usecases/mount/UC_delete_from_rdb.md [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+## Scenarios
+
+- **Precondition:**
+
+  - The administrators have a working installation of Elektra and a tool to access the KDB (key database) on their system.
+  - The database is mounted into the KDB. (see [Use Case: Mount a data source](./UC_mount.md))
+
+- **Main success scenario:**
+
+  - The administrators query a value from the mounted database by giving the unique keyname. (see [Use Case: Read a configuration value from a file](./UC_read_rdb.md))
```
Comment (flo91) [inline doc/usecases/mount/UC_delete_from_rdb.md:20]:
```suggestion
  - The administrators query a value from the mounted database by giving the unique keyname. (see [Use Case: Read a configuration value from a relational database](./UC_read_rdb.md))
```

File: doc/usecases/mount/UC_edit_rdb.md [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+## Scenarios
+
+- **Precondition:**
+
+  - The administrators have a working installation of Elektra and a tool to access the KDB (key database) on their system.
+  - The database is mounted into the KDB. (see [Use Case: Mount a data source](./UC_mount.md))
+
+- **Main success scenario:**
+
+  - The administrators query a value from the mounted database by giving the unique keyname. (see [Use Case: Read a configuration value from a file](./UC_read_rdb.md))
```
Comment (flo91) [inline doc/usecases/mount/UC_edit_rdb.md:20]:
```suggestion
  - The administrators query a value from the mounted database by giving the unique keyname. (see [Use Case: Read a configuration value from a relational database](./UC_read_rdb.md))
```

File: doc/usecases/mount/UC_edit_rdb.md (hunk 12 above)
Comment (flo91) [inline doc/usecases/mount/UC_edit_rdb.md:31]:
```suggestion
    - Elektra creates a new tuple with the given key and value and writes it to the database.
```

Comment (flo91) [reply doc/usecases/mount/UC_edit_rdb.md:31]: Thank you for the remark! I changed the scenario.

**Pull request closed without merging**